pylabrobot/pumps/pumparray.py

```diff
@@ -0,0 +1,142 @@
+    """
+
+    if isinstance(use_channels, int):
+      use_channels = [use_channels]
+    if isinstance(num_revolutions, float):
+      num_revolutions = [num_revolutions] * len(use_channels)
+    await self.backend.run_revolutions(num_revolutions=num_revolutions, use_channels=use_channels)
+
+  async def run_continuously(self, speed: Union[float, int, List[float], List[int]],
+                             use_channels: Union[int, List[int]]):
+    """ Run the specified channels at the specified speeds.
+
+    Args:
+      speed: speed in rpm/pump-specific units.
+      use_channels: pump array channels to run.
+    """
+
+    if isinstance(use_channels, list) and len(set(use_channels)) != len(use_channels):
+      raise ValueError("Channels in use channels must be unique.")
+    if isinstance(use_channels, int):
+      use_channels = [use_channels]
+    if isinstance(speed, (float, int)):
+      speed = [speed] * len(use_channels)
+    if isinstance(speed[0], int):
+      speed = [float(x) for x in speed]
+    if len(speed) != len(use_channels):
+      raise ValueError("Speed and use_channels must be the same length.")
+    if any(channel < 0 for channel in use_channels):
+      raise ValueError("Channels in use channels must be positive.")
+    await self.backend.run_continuously(speed=speed,  # type: ignore[arg-type]
+                                        use_channels=use_channels)
+
+  async def run_for_duration(self, speed: Union[float, int, List[float], List[int]],
+                             use_channels: Union[int, List[int]],
+                             duration: Union[float, int]):
+    """ Run the specified channels at the specified speeds for the specified duration.
+
+    Args:
+      speed: speed in rpm/pump-specific units.
+      use_channels: pump array channels to run.
+      duration: duration to run pumps (seconds).
+    """
+
+    if duration < 0:
+      raise ValueError("Duration must be positive.")
+    await self.run_continuously(speed=speed, use_channels=use_channels)
+    await asyncio.sleep(duration)
+    await self.run_continuously(speed=0, use_channels=use_channels)
+
+  async def pump_volume(self, speed: Union[float, int, List[float], List[int]],
+                        use_channels: Union[int, List[int]],
+                        volume: Union[float, int, List[float], List[int]],
+                        calibration_units: Literal["duration", "revolutions"]):
+    """ Run the specified channels at the specified speeds for the specified volume. Note that this
+    function requires the pump to be calibrated at the input speed.
+
+    Args:
+      speed: speed in rpm/pump-specific units. use_channels: pump array channels to run using
+        0-index. volume: volume to pump.
+    calibration_units: units of calibration. Volume per seconds ("duration") or volume per
+      revolution ("revolutions").
+
+    Raises:
+      TypeError: if the pump is not calibrated.
+    """
```

Comment:
Since calibration is specified on the class, I think `calibration_units` should be as well. Perhaps it is even better to store the calibration units on the PumpCalibration object since they'd always be used together.